pkg/private/zip/build_zip.py

```diff
@@ -68,53 +69,145 @@ def parse_date(ts):
+    if entry_type == manifest.ENTRY_IS_DIR and not dst_path.endswith('/'):
+      dst_path += '/'
+    entry_info = self.make_zipinfo(path=dst_path, mode=mode)
+
+    if entry_type == manifest.ENTRY_IS_FILE:
+      entry_info.compress_type = zipfile.ZIP_DEFLATED
+      with open(src, 'rb') as src:
+        self.zip_file.writestr(entry_info, src.read())
+    elif entry_type == manifest.ENTRY_IS_DIR:
+      entry_info.compress_type = zipfile.ZIP_STORED
+      # Set directory bits
+      entry_info.external_attr |= (UNIX_DIR_BIT << 16) | MSDOS_DIR_BIT
+      self.zip_file.writestr(entry_info, '')
+    elif entry_type == manifest.ENTRY_IS_LINK:
+      entry_info.compress_type = zipfile.ZIP_STORED
+      # Set directory bits
+      entry_info.external_attr |= (UNIX_SYMLINK_BIT << 16)
+      self.zip_file.writestr(entry_info, src)
+    elif entry_type == manifest.ENTRY_IS_TREE:
+      self.add_tree(src, dst_path, mode)
+    elif entry.entry_type == manifest.ENTRY_IS_EMPTY_FILE:
+      entry_info.compress_type = zipfile.ZIP_DEFLATED
+      self.zip_file.writestr(entry_info, '')
+    else:
+      raise Exception('Unknown type for manifest entry:', entry)
+
+  def add_tree(self, tree_top: str, destpath: str, mode: int):
+    """Add a tree artifact to the zip file.
+
+    Args:
+      tree_top: the top of the tree to add
+      destpath: the path under which to place the files
+      mode: if not None, file mode to apply to all files
+    """
+
+    # We expect /-style paths.
+    tree_top = os.path.normpath(tree_top).replace(os.path.sep, '/')
+
+    # Again, we expect /-style paths.
+    dest = destpath.strip('/')  # redundant, dests should never have / here
+    dest = os.path.normpath(dest).replace(os.path.sep, '/')
+    # paths should not have a leading ./
+    dest = '' if dest == '.' else dest + '/'
+
+    to_write = {}
+    for root, dirs, files in os.walk(tree_top):
+      # While `tree_top` uses '/' as a path separator, results returned by
+      # `os.walk` and `os.path.join` on Windows may not.
+      root = os.path.normpath(root).replace(os.path.sep, '/')
+
+      dirs = sorted(dirs)
+      rel_path_from_top = root[len(tree_top):].lstrip('/')
+      if rel_path_from_top:
+        dest_dir = dest + rel_path_from_top + '/'
+      else:
+        dest_dir = dest
+      for file in sorted(files):
+        to_write[dest_dir + file] = root + '/' + file
+
+    for path in sorted(to_write.keys()):
+      content_path = to_write[path]
+      # If mode is unspecified, derive the mode from the file's mode.
+      if mode is None:
+        f_mode = 0o755 if os.access(content_path, os.X_OK) else 0o644
+      else:
+        f_mode = mode
+      entry_info = self.make_zipinfo(path=path, mode=f_mode)
+      entry_info.compress_type = zipfile.ZIP_DEFLATED
+      if not content_path:
+        self.zip_file.writestr(entry_info, '')
+      else:
+        with open(content_path, 'rb') as src:
+          self.zip_file.writestr(entry_info, src.read())
```

Comment:
Would `add_tree` takes care the cases when content_path points to an empty file or a link?